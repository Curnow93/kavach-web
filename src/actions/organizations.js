import axios from 'axios';
import {
  ADD_ORGANIZATION,
  ADD_ORGANIZATIONS,
  SET_ORGANIZATIONS_LOADING,
  RESET_ORGANIZATIONS,
  ORGANIZATIONS_API,
  SET_SELECTED_ORGANIZATION,
} from '../constants/organizations';

export const getOrganizations = () => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .get(ORGANIZATIONS_API + '/my')
      .then((response) => {
        dispatch(addOrganizationsList(response.data));
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const getOrganization = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .get(ORGANIZATIONS_API + '/' + id)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const addOrganization = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .post(ORGANIZATIONS_API, data)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(setSelectedOrganization(response.data.id));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const updateOrganization = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .put(ORGANIZATIONS_API + '/' + data.id, data)
      .then((response) => {
        dispatch(getOrganizationByID(response.data));
        dispatch(stopOrganizationsLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const deleteOrganization = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganizations());
    return axios
      .delete(ORGANIZATIONS_API + '/' + id)
      .then(() => {
        dispatch(resetOrganizations());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const setSelectedOrganization = (id) => ({
  type: SET_SELECTED_ORGANIZATION,
  payload: id,
});

const loadingOrganizations = () => ({
  type: SET_ORGANIZATIONS_LOADING,
  payload: true,
});

const stopOrganizationsLoading = () => ({
  type: SET_ORGANIZATIONS_LOADING,
  payload: false,
});

const getOrganizationByID = (data) => ({
  type: ADD_ORGANIZATION,
  payload: data,
});

const addOrganizationsList = (data) => ({
  type: ADD_ORGANIZATIONS,
  payload: data,
});

const resetOrganizations = () => ({
  type: RESET_ORGANIZATIONS,
});
