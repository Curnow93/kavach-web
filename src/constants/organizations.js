//Actions
export const ADD_ORGANIZATION = 'ADD_ORGANIZATION';
export const ADD_ORGANIZATIONS = 'ADD_ORGANIZATIONS';
export const RESET_ORGANIZATIONS = 'RESET_ORGANIZATIONS';
export const SET_ORGANIZATIONS_LOADING = 'SET_ORGANIZATIONS_LOADING';
export const SET_SELECTED_ORGANIZATION = 'SET_SELECTED_ORGANIZATION';
//API
export const ORGANIZATIONS_API = '/organizations';
