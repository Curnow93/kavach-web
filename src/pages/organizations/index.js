import React from 'react';
import OrganizationUsers from './components/users';
import { Card, List, Button, Form, Input, Popconfirm, Divider, Modal } from 'antd';
import { DeleteOutlined, EditOutlined, CloseOutlined } from '@ant-design/icons';

function Organizations() {
  const [organizations, setOrganizations] = React.useState([]);
  const [edit, setEdit] = React.useState(null);
  const [showModal, setShowModal] = React.useState(false);
  const [loading, setLoading] = React.useState(false);

  React.useEffect(() => {
    fetch(process.env.REACT_APP_API_URL + '/organizations/my')
      .then((res) => {
        if (res.status === 200) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => setOrganizations(res))
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const updateOrg = (values, id) => {
    fetch(process.env.REACT_APP_API_URL + '/organizations/' + id, {
      method: 'PUT',
      body: JSON.stringify(values),
    })
      .then((res) => {
        if (res.status === 200) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        var updateIndex = organizations.findIndex((item) => item.id === id);
        if (updateIndex > -1) {
          var newData = organizations;
          newData[updateIndex] = res;
          setOrganizations(newData);
          setEdit(null);
        }
      })
      .catch((err) => {
        console.log(err);
      });
  };

  const deleteOrg = (id) => {
    setLoading(true);
    fetch(process.env.REACT_APP_API_URL + '/organizations/' + id, {
      method: 'DELETE',
    })
      .then((res) => {
        if (res.status === 200) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        var deleteIndex = organizations.findIndex((item) => item.id === id);
        if (deleteIndex > -1) {
          var newData = organizations;
          newData.splice(deleteIndex, 1);

          setOrganizations(newData);
        }
      })
      .catch((err) => {
        console.log(err);
      });
    setLoading(false);
  };

  const addOrg = (values) => {
    setLoading(true);
    fetch(process.env.REACT_APP_API_URL + '/organizations/', {
      method: 'POST',
      body: JSON.stringify(values),
    })
      .then((res) => {
        if (res.status === 201) {
          return res.json();
        } else {
          throw new Error(res.status);
        }
      })
      .then((res) => {
        var newData = organizations;
        newData.push(res);
        setOrganizations(newData);
      })
      .catch((err) => {
        console.log(err);
      });
    setLoading(false);
    setShowModal(false);
  };

  return (
    <div className="content">
      <div id="add-organization">
        <Button type="primary" onClick={() => setShowModal(true)}>
          Add Organization
        </Button>
        <Modal
          title="Add Organization"
          visible={showModal}
          onCancel={() => setShowModal(false)}
          footer={[
            <Button key="back" onClick={() => setShowModal(false)}>
              Cancel
            </Button>,
            <Button form="organization_new" key="submit" htmlType="submit" type="primary">
              Submit
            </Button>,
          ]}
        >
          <Form name="organization_new" onFinish={(values) => addOrg(values)}>
            <Form.Item name="title" label="Title">
              <Input placeholder="title" />
            </Form.Item>
            <Form.Item name="slug" label="Slug">
              <Input placeholder="slug" />
            </Form.Item>
          </Form>
        </Modal>
      </div>
      <div id="organization-list">
        <List
          loading={loading}
          grid={{ gutter: 16, column: 3 }}
          dataSource={organizations}
          itemLayout="vertical"
          renderItem={(item) => (
            <List.Item key={item.id}>
              <Card
                title={
                  item.id === edit ? (
                    <Form
                      name="organization_edit"
                      layout="inline"
                      onFinish={(values) => updateOrg(values, item.id)}
                      initialValues={{
                        title: item.title,
                      }}
                    >
                      <Form.Item name="title">
                        <Input placeholder="title" />
                      </Form.Item>
                      <Form.Item>
                        <Button form="organization_edit" type="primary" htmlType="submit" block>
                          Save
                        </Button>
                      </Form.Item>
                    </Form>
                  ) : (
                    item.title
                  )
                }
                extra={
                  item.permission.role === 'owner' ? (
                    <div>
                      <Button
                        icon={item.id === edit ? <CloseOutlined /> : <EditOutlined />}
                        onClick={() => (item.id === edit ? setEdit(null) : setEdit(item.id))}
                      />
                      <Divider type="vertical" />
                      <Popconfirm title="Sure to delete?" onConfirm={() => deleteOrg(item.id)}>
                        <Button icon={<DeleteOutlined />} />
                      </Popconfirm>
                    </div>
                  ) : null
                }
              >
                <OrganizationUsers organization={item} />
              </Card>
            </List.Item>
          )}
        />
      </div>
    </div>
  );
}

export default Organizations;
