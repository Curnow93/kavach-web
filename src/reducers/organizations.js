import {
  ADD_ORGANIZATION,
  ADD_ORGANIZATIONS,
  SET_ORGANIZATIONS_LOADING,
  RESET_ORGANIZATIONS,
  SET_SELECTED_ORGANIZATION,
} from '../constants/organizations';

const initialState = {
  ids: [],
  details: {},
  loading: true,
  selected: 0,
};

export default function tagsReducer(state = initialState, action = {}) {
  switch (action.type) {
    case RESET_ORGANIZATIONS:
      return {
        ...state,
        ids: [],
        details: {},
        loading: true,
        selected: 0,
      };
    case SET_ORGANIZATIONS_LOADING:
      return {
        ...state,
        loading: action.payload,
      };
    case ADD_ORGANIZATIONS:
      return {
        ...state,
        ids: action.payload.map((item) => item.id),
        details: action.payload.reduce((obj, item) => Object.assign(obj, { [item.id]: item }), {}),
        selected: action.payload[0].id,
      };
    case ADD_ORGANIZATION:
      return {
        ...state,
        ids: state.ids.includes(action.payload.id)
          ? state.ids
          : state.ids.concat([action.payload.id]),
        details: {
          ...state.details,
          [action.payload.id]: action.payload,
        },
      };
    case SET_SELECTED_ORGANIZATION:
      return {
        ...state,
        selected: action.payload,
      };
    default:
      return state;
  }
}
