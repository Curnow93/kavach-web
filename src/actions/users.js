import axios from 'axios';
import { ADD_USERS, SET_USERS_LOADING, RESET_USERS, USERS_API } from '../constants/users';

export const getUsers = () => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .get(USERS_API + '/' + getState().organizations.selected + '/users')
      .then((response) => {
        dispatch(addUsersList(response.data));
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const addUser = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .post(USERS_API + '/' + getState().organizations.selected + '/users', data)
      .then(() => {
        dispatch(resetUsers());
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

export const deleteUser = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .delete(USERS_API + '/' + getState().organizations.selected + '/users/' + id)
      .then(() => {
        dispatch(resetUsers());
        dispatch(stopUsersLoading());
      })
      .catch((error) => {
        console.log(error.message);
      });
  };
};

const loadingUsers = () => ({
  type: SET_USERS_LOADING,
  payload: true,
});

const stopUsersLoading = () => ({
  type: SET_USERS_LOADING,
  payload: false,
});

const addUsersList = (data) => ({
  type: ADD_USERS,
  payload: data,
});

const resetUsers = () => ({
  type: RESET_USERS,
});
