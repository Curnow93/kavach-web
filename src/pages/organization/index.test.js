import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';

import '../../matchMedia.mock';
import OrganizationCreate from './index';
import { addOrganization } from '../../actions/organizations';

jest.mock('../../actions/organizations', () => ({
  addOrganization: jest.fn(),
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

describe('Organizations index component', () => {
  const mockedDispatch = jest.fn();
  mockedDispatch.mockReturnValue(Promise.resolve());
  useDispatch.mockReturnValue(mockedDispatch);

  const push = jest.fn();
  useHistory.mockReturnValue({ push });
  describe('snapshot testing', () => {
    it('should render the component', () => {
      let component;
      act(() => {
        component = shallow(<OrganizationCreate />);
      });
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    beforeEach(() => {
      act(() => {
        wrapper = mount(<OrganizationCreate />);
      });
    });
    afterEach(() => {
      wrapper.unmount();
    });
    it('should submit form with data', (done) => {
      act(() => {
        wrapper
          .find('FormItem')
          .at(0)
          .find('Input')
          .simulate('change', { target: { value: 'title' } });
        wrapper
          .find('FormItem')
          .at(1)
          .find('TextArea')
          .at(0)
          .simulate('change', { target: { value: 'description' } });

        const submitButtom = wrapper.find('Button').at(0);
        submitButtom.simulate('submit');
      });
      wrapper.update();

      setTimeout(() => {
        expect(addOrganization).toHaveBeenCalledTimes(1);
        expect(addOrganization).toHaveBeenCalledWith({
          title: 'title',
          description: 'description',
        });
        expect(push).toHaveBeenCalledWith('/settings');
        done();
      });
    });
  });
});
