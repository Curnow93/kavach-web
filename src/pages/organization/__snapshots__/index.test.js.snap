// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organizations index component snapshot testing should render the component 1`] = `ShallowWrapper {}`;
